sdlib/d/sync/sharedlock.d

```diff
@@ -0,0 +1,558 @@
+	 *    shared locks should wait.
+	 * 3. When all shared locks are released, the exclusive lock is taken.
+	 * 4. When the exclusive lock is released, shared locks can again be taken.
+	 *
+	 * To accomplish this, we will start one exclusive thread and several
+	 * sharing threads. We will issue commands for each action, and then
+	 * wait for the desired state, then we measure the lock state and
+	 * ensure everything is correct.
+	 */
+	static runThread(void* delegate() dg) {
+		static struct Delegate {
+			void* ctx;
+			void* function(void*) fun;
+		}
+
+		auto x = *(cast(Delegate*) &dg);
+
+		import core.stdc.pthread;
+		pthread_t tid;
+		auto r = pthread_create(&tid, null, x.fun, x.ctx);
+		assert(r == 0, "Failed to create thread!");
+
+		return tid;
+	}
+
+	shared SharedLock lock;
+
+	import d.sync.atomic;
+	import sys.posix.sched;
+	shared Atomic!uint exclusiveLockState;
+	shared Atomic!uint sharedLockCount;
+	shared Atomic!uint attemptedSharedLockCount;
+	shared Atomic!uint desiredSharedLocks;
+
+	void* exclusiveLocker() {
+		while (true) {
+			uint cur = exclusiveLockState.load();
+			auto s = cur;
+			assert((s & 1) == 0);
+			while (s == cur) {
+				sched_yield();
+				s = exclusiveLockState.load();
+			}
+
+			scope(exit) exclusiveLockState.fetchAdd(1);
+
+			switch (s) {
+				case 1: // Lock
+					lock.exclusiveLock();
+					break;
+				case 3: // Unlock
+					lock.exclusiveUnlock();
+					break;
+				default: // Exit
+					return null;
+			}
+		}
+	}
+
+	void* sharedLocker() {
+		uint id = sharedLockCount.fetchAdd(1);
+		bool lockHeld = false;
+		while (true) {
+			uint cmd = desiredSharedLocks.load();
+			if (cmd == uint.max) {
+				return null;
+			}
+
+			if (cmd > id && !lockHeld) {
+				attemptedSharedLockCount.fetchAdd(1);
+				lock.sharedLock();
+				attemptedSharedLockCount.fetchSub(1);
+				lockHeld = true;
+				sharedLockCount.fetchAdd(1);
+			} else if (cmd <= id && lockHeld) {
+				lock.sharedUnlock();
+				lockHeld = false;
+				sharedLockCount.fetchSub(1);
+			} else {
+				sched_yield();
+			}
+		}
+	}
+
+	enum SharerThreads = 5;
+
+	import core.stdc.pthread;
+	pthread_t[SharerThreads + 1] tids;
+	tids[0] = runThread(exclusiveLocker);
+	foreach (i; 1 .. tids.length) {
+		tids[i] = runThread(sharedLocker);
+	}
+
+	void waitForShared(uint count, uint attempts) {
+		while (sharedLockCount.load() != count
+			       && attemptedSharedLockCount.load() != attempts) {
+			sched_yield();
+		}
+
+		// Sleep a tiny bit (10ms) to ensure a steady state
+		import core.stdc.unistd;
+		usleep(10 * 1000);
+
+		assert(sharedLockCount.load() == count,
+		       "Unexpected change of shared locks!");
+		assert(attemptedSharedLockCount.load() == attempts,
+		       "Unexpected change of shared attempts!");
+	}
+
+	// Wait for all ids to be assigned.
+	waitForShared(SharerThreads, 0);
+	assert(lock.count == 0, "Expected count not correct!");
+	assert(!lock.mutex.isHeld(), "Mutex is held!");
+
+	sharedLockCount.store(0);
```

Comment:
This is a time bomb (and I suspect the reason why you felt the need to ass the sleep above?). Just do not reuse the same state for two things.

Reply:
It is not the reason for the sleep, it's something where I needed to avoid a situation where multiple threads are forever incrementing and decrementing a counter because it's going around the desired steady-state. I wanted to know that when the steady state was reached, it should be done. So I assign an id to each sharer thread so it can be deterministic how many threads should try to lock.

This was because I didn't add a second variable to assign the ids, since this isn't yet being used. I'll switch it out.

Reply:
fixed.